File: tests/httpShowPvd.js
#!/usr/bin/env node

/*
	Copyright 2017 Cisco

	Licensed under the Apache License, Version 2.0 (the "License");
	you may not use this file except in compliance with the License.
	You may obtain a copy of the License at

		http://www.apache.org/licenses/LICENSE-2.0

	Unless required by applicable law or agreed to in writing, software
	distributed under the License is distributed on an "AS IS" BASIS,
	WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
	See the License for the specific language governing permissions and
	limitations under the License.
*/
/*
 * Simple http server :
 * http://localhost:8200 => displays the list of pvds
 * http://localhost:8200/<pvd> => displays the attributes for <pvd>
 */
var http = require("http");
var pvdd = require("pvdd");

var cnx = pvdd.connect({ "autoReconnect" : true });
var allPvd = {};

cnx.on("connect", function() {
	cnx.subscribeAttribute("*");
	cnx.subscribeNotifications();
	cnx.getList();
});

cnx.on("error", function(err) {});

cnx.on("pvdList", function(pvdList) {
	pvdList.forEach(function(pvd) {
		cnx.getAttributes(pvd);
	});
	allPvd = {};
});

cnx.on("pvdAttributes", function(pvd, attrs) {
	allPvd[pvd] = attrs;
});

function pvd2str(pvd) {
	return JSON.stringify(allPvd[pvd] || {}, null, '\t');
}

var server = http.createServer(function(req, res) {
	h = {};
	pvd = req.url.slice(1);
	if (pvd == null || pvd == "") {
		s = "<!DOCTYPE html><html><head>" +
		    "<title>Provisionning domains</title>" +
		    "</head><body>\n";
		Object.keys(allPvd).forEach(function(pvd) {
			s += "<a href=http://" + req.headers.host + "/" + pvd +
			     " title='" + pvd2str(pvd).replace("'", "&#39;") +
			     "'>" + pvd + "</a><br>\n";
		});
		s += "</body></html>";
	}
	else {
		h = { 'Content-Type': 'application/json' };
		s = pvd2str(pvd);
	}
	res.writeHead(200, h);
	res.end(s);
});

var HttpPort = 8200;

server.listen(HttpPort, "::");

File: bindings/python-3/httpShowPvd.py
#!/usr/bin/env python3

#
#	Copyright 2017 Cisco
#
#	Licensed under the Apache License, Version 2.0 (the "License");
#	you may not use this file except in compliance with the License.
#	You may obtain a copy of the License at
#
#		http://www.apache.org/licenses/LICENSE-2.0
#
#	Unless required by applicable law or agreed to in writing, software
#	distributed under the License is distributed on an "AS IS" BASIS,
#	WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#	See the License for the specific language governing permissions and
#	limitations under the License.
#

# Simple http server :
# http://localhost:8100 => displays the list of pvds
# http://localhost:8100/<pvd> => displays the attributes for <pvd>
from http import server
import pvdd
import json

PORT = 8100

class MyHandler(server.BaseHTTPRequestHandler):
    protocol_version = "HTTP/1.0"
    def do_GET(self):
        host = self.headers["Host"] if "Host" in self.headers else "localhost:" + str(PORT)
        self.send_response(200)
        if self.path == '/':
            s = "<!DOCTYPE html><html><head>" +\
                "<title>Provisionning domains</title>" +\
                "</head><body>\n"
            for pvd in sorted(allPvd):
                s = s + "<a href=http://" + host + "/" + pvd +\
                        " title='" + pvd2str(pvd).replace("'", "&#39;") +\
                        "'>" + pvd + "</a><br>\n"
            s = s + "</body></html>"
            self.send_header("Content-Type", "text/html")
        else:
            self.send_header("Content-Type", "application/json")
            s = pvd2str(self.path.strip("/"))

        self.end_headers()
        self.wfile.write(s.encode("utf-8"))

allPvd = {}

def handleConnect(pvdd):
    allPvd = {}
    pvdd.subscribeNotifications()
    pvdd.subscribeAttribute("*")
    pvdd.getList()

def handlePvdList(pvdd, pvdList):
    for pvd in pvdList:
        pvdd.getAttributes(pvd)

def handlePvdAttributes(pvdd, pvd, attrs):
    allPvd[pvd] = attrs

def pvd2str(pvd):
        return json.dumps(allPvd[pvd], indent='\t') if pvd in allPvd else "{}"

pvdd = pvdd.pvdd()
pvdd.on("connect", handleConnect)
pvdd.on("pvdList", handlePvdList)
pvdd.on("pvdAttributes", handlePvdAttributes)
pvdd.connect(autoReconnect = True, verbose = False)

httpd = server.HTTPServer(("", PORT), MyHandler)
sa = httpd.socket.getsockname()
print("Serving HTTP on", sa[0], "port", sa[1], "...")
try:
    httpd.serve_forever()
except KeyboardInterrupt:
    print("\nKeyboard interrupt received, exiting.")
    httpd.server_close()

pvdd.leave()

